mlflow/models/utils.py

```diff
@@ -264,3 +276,301 @@ def plot_lines(data_series, xlabel, ylabel, legend_loc=None, line_kwargs=None):
+    and type specified in model's input schema.
+    """
+    if not input_schema.is_tensor_spec():
+        if isinstance(pfInput, (list, np.ndarray, dict, pd.Series)):
+            try:
+                pfInput = pd.DataFrame(pfInput)
+            except Exception as e:
+                raise MlflowException(
+                    "This model contains a column-based signature, which suggests a DataFrame"
+                    " input. There was an error casting the input data to a DataFrame:"
+                    " {0}".format(str(e))
+                )
+        if not isinstance(pfInput, pd.DataFrame):
+            raise MlflowException(
+                "Expected input to be DataFrame or list. Found: %s" % type(pfInput).__name__
+            )
+
+    if input_schema.has_input_names():
+        # make sure there are no missing columns
+        input_names = input_schema.input_names()
+        expected_cols = set(input_names)
+        actual_cols = set()
+        if len(expected_cols) == 1 and isinstance(pfInput, np.ndarray):
+            # for schemas with a single column, match input with column
+            pfInput = {input_names[0]: pfInput}
+            actual_cols = expected_cols
+        elif isinstance(pfInput, pd.DataFrame):
+            actual_cols = set(pfInput.columns)
+        elif isinstance(pfInput, dict):
+            actual_cols = set(pfInput.keys())
+        missing_cols = expected_cols - actual_cols
+        extra_cols = actual_cols - expected_cols
+        # Preserve order from the original columns, since missing/extra columns are likely to
+        # be in same order.
+        missing_cols = [c for c in input_names if c in missing_cols]
+        extra_cols = [c for c in actual_cols if c in extra_cols]
+        if missing_cols:
+            message = "Model is missing inputs {0}.".format(missing_cols)
+            if extra_cols:
+                message += " Note that there were extra inputs: {0}".format(extra_cols)
+            raise MlflowException(message)
+    elif not input_schema.is_tensor_spec():
+        # The model signature does not specify column names => we can only verify column count.
+        num_actual_columns = len(pfInput.columns)
+        if num_actual_columns < len(input_schema.inputs):
+            raise MlflowException(
+                "Model inference is missing inputs. The model signature declares "
+                "{0} inputs  but the provided value only has "
+                "{1} inputs. Note: the inputs were not named in the signature so we can "
+                "only verify their count.".format(len(input_schema.inputs), num_actual_columns)
+            )
+
+    return (
+        _enforce_tensor_schema(pfInput, input_schema)
+        if input_schema.is_tensor_spec()
+        else _enforce_col_schema(pfInput, input_schema)
+    )
+
+
+def validate_schema(data: DataInputType, expected_schema: Schema) -> DataInputType:
```

Comment:
Following from https://github.com/mlflow/mlflow/pull/6494/files#r950650995, should we change this to:

```suggestion
def validate_schema(data: PyFuncInput, expected_schema: Schema) -> DataInputType:
```

Since `_enforce_schema` doesn't support pandas series and validation of output schemas doesn't appear to be critically important?

Reply:
I actually added support of pandas series, as it can be directly turned into a pandas DataFrame and does the following validation. And in fact we could accept pandas.Series as a dataset input.

Reply:
I think it is nice to make `PyFuncInput` support `pd.Series` type. @dbczumar Thoughts ?

But, @serena-ruan , your definition `DataInputType = Union[PyFuncInput, PyFuncOutput]` looks weird. We can directly reuse `PyFuncInput` here, if we add `pd.Series` into `PyFuncInput`

Reply:
Ahh yes, will drop DataInputType.